File: Gaussian_Metropolis.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

int main(void){
  int niter=100;
  double step_size=0.5e0;

  srand((unsigned)time(NULL)); 

  /*********************************/
  /* Set the initial configuration */
  /*********************************/      
  double x=0e0;
  int naccept=0;
  /*************/
  /* Main loop */
  /*************/
  for(int iter=1;iter<niter+1;iter++){
    double backup_x=x;    
    double action_init=0.5e0*x*x;
    
    double dx = (double)rand()/RAND_MAX;
    dx=(dx-0.5e0)*step_size*2e0;
    x=x+dx;
    
    double action_fin=0.5e0*x*x;
    /*******************/
    /* Metropolis test */
    /*******************/
    double metropolis = (double)rand()/RAND_MAX;    
    if(exp(action_init-action_fin) > metropolis)
      /* accept */
      naccept=naccept+1;
    else 
      /* reject */
      x=backup_x;
    /***************/
    /* data output */
    /***************/
    printf("%.10f   %f\n",x,(double)naccept/iter);}
}

File: Gaussian_Metropolis.py
import numpy as np
from matplotlib import pyplot as plt
# seed of random numbers is the system time by default

NITER = 1000000
STEP_SIZE = 0.5
#####################################
### Set the initial configuration ###
#####################################
x = 0
naccept = 0

## data for plot
data_for_plot = []

#################
### Main Loop ###
#################
for iter in range(NITER):
    backup_x = x
    action_init = 0.5*x*x
    dx = np.random.uniform(-STEP_SIZE, STEP_SIZE)
    #x = x + dx
    x += dx
    action_fin = 0.5*x*x
    #######################
    ### Metropolis test ###
    #######################
    metropolis = np.random.uniform(0,1)
    if(np.exp(action_init-action_fin) > metropolis):#accept
      #naccept = naccept+1
      naccept += 1
    else:#reject
      x = backup_x
    print(x,naccept/(iter+1))

    ## save configuration for plot
    data_for_plot.append(x)

###########################################################
### plot the histogram and compare with analytic result ###
###########################################################
def gauss(x):
    return 1.0/np.sqrt(2.0*np.pi)*np.exp( -x*x/2.0 )

y=np.arange(-4.0,4.0,0.1)
plt.xlabel("$x$")
plt.ylabel("$P(x)$")
plt.plot(y,gauss(y), label="exact")
plt.hist(data_for_plot,bins=80,density=True, label="MCMC")
plt.legend()
plt.show()
